src/c/backend/task.cpp

```diff
@@ -301,3 +305,54 @@ void InnerTask::set_complete() {
 }
 
 bool InnerTask::get_complete() { return this->complete.load(); }
+
+
+// TODO(hc): The current Parla exploits two types of resources,
+//           memory and vcus. Later, this can be extended with
+//           a map.
+void InnerTask::add_device_req(Device* dev_ptr, MemorySzTy mem_sz, VCUTy num_vcus) {
+  DeviceResources res_req = {mem_sz, num_vcus};
+  DeviceRequirement* dev_req = new DeviceRequirement(dev_ptr, res_req);
+  if (req_addition_mode_ == SingleDevAdd) {
+    dev_res_reqs_.AppendDeviceRequirementOption(dev_req);
+  } else if (req_addition_mode_ % 2 == 0) { /* Architecture requirement */
+    tmp_arch_req_->AppendDeviceRequirementOption(dev_req);
+  } else if (req_addition_mode_ == MultiDevAdd) {
+    tmp_multdev_reqs_->AppendDeviceRequirement(dev_req);
+  }
+}
+
+void InnerTask::begin_arch_req_addition() {
```

Comment:
What's the benefit of this pattern (just curious)?

Reply:
Basically this pattern was came from the lack of idea. You can understand this as a transaction-based method. We need to handle nested/recursive patterns even though this is at most 2 levels. I have hierarchical vectors/classes and the lower level vectors containing requirements should be pushed to the higher level. For example, architecture requirement consisting of multiple device requirements could be pushed into multi-device requirement vector or the topmost vector. I can distinguish the levels easily with those functions.